// @kata-kit/common/lib/index.es.js
import styled, { keyframes } from 'styled-components';
import { createElement, Component, createRef } from 'react';
import createFocusTrap from 'focus-trap';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}

function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var Board = (function (_super) {
    __extends(Board, _super);
    function Board() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Board.prototype.render = function () {
        var _a = this.props, className = _a.className, children = _a.children;
        return (createElement(Root, { className: className },
            this.renderHeading(),
            createElement(Body, null, children),
            this.renderFooter()));
    };
    Board.prototype.renderHeading = function () {
        if (this.props.headingChildren) {
            return createElement(Heading, null, this.props.headingChildren);
        }
        return null;
    };
    Board.prototype.renderFooter = function () {
        if (this.props.footerChildren) {
            return createElement(Footer, null, this.props.footerChildren);
        }
        return null;
    };
    return Board;
}(Component));
var Root = styled('div')(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  background-color: #fff /* $white */;\n  border-radius: 6px /* $border-radius-medium */;\n  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.25) /* $layer-100-shadow */;\n"], ["\n  background-color: #fff /* $white */;\n  border-radius: 6px /* $border-radius-medium */;\n  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.25) /* $layer-100-shadow */;\n"])));
var Heading = styled('div')(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  padding: 24px 40px;\n  border-bottom: 1px solid #e2e6e8 /* $gray-30 */;\n"], ["\n  padding: 24px 40px;\n  border-bottom: 1px solid #e2e6e8 /* $gray-30 */;\n"])));
var Body = styled('div')(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  padding: 24px 40px;\n"], ["\n  padding: 24px 40px;\n"])));
var Footer = styled('div')(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  padding: 24px 40px;\n  border-top: 1px solid #e2e6e8 /* $gray-30 */;\n"], ["\n  padding: 24px 40px;\n  border-top: 1px solid #e2e6e8 /* $gray-30 */;\n"])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;

var EmptyMessage = (function (_super) {
    __extends(EmptyMessage, _super);
    function EmptyMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EmptyMessage.prototype.render = function () {
        return (createElement(Wrapper, null,
            createElement("img", { src: this.props.image, alt: "Empty Message" }),
            createElement(Title, { "data-testid": "EmptyMessage-title" }, this.props.title),
            createElement(Description, null, this.props.children)));
    };
    EmptyMessage.defaultProps = {
        image: require('@kata-kit/assets/images/form-empty.svg'),
        title: 'Empty'
    };
    return EmptyMessage;
}(Component));
var FadeIn = keyframes(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n    transform: translateY(5%);\n  }\n  to {\n    opacity: 1;\n    transform: translateY(0%);\n  }\n"], ["\n  from {\n    opacity: 0;\n    transform: translateY(5%);\n  }\n  to {\n    opacity: 1;\n    transform: translateY(0%);\n  }\n"])));
var Title = styled('h3')(templateObject_2$1 || (templateObject_2$1 = __makeTemplateObject(["\n  margin: 24px 0 8px;\n  text-align: center;\n  color: #24282d;\n"], ["\n  margin: 24px 0 8px;\n  text-align: center;\n  color: #24282d;\n"])));
var Description = styled('p')(templateObject_3$1 || (templateObject_3$1 = __makeTemplateObject(["\n  margin-top: 8px;\n  margin-bottom: 0;\n  text-align: center;\n  color: #484c4f;\n"], ["\n  margin-top: 8px;\n  margin-bottom: 0;\n  text-align: center;\n  color: #484c4f;\n"])));
var Wrapper = styled('div')(templateObject_4$1 || (templateObject_4$1 = __makeTemplateObject(["\n  animation: ", " 0.75s ease;\n  max-width: 320px;\n  margin: 48px auto;\n  text-align: center;\n\n  ", " {\n    font-weight: 500;\n  }\n\n  ", " {\n    font-weight: 300;\n  }\n"], ["\n  animation: ", " 0.75s ease;\n  max-width: 320px;\n  margin: 48px auto;\n  text-align: center;\n\n  ", " {\n    font-weight: 500;\n  }\n\n  ", " {\n    font-weight: 300;\n  }\n"])), FadeIn, Title, Description);
var templateObject_1$1, templateObject_2$1, templateObject_3$1, templateObject_4$1;

var FocusTrap = (function (_super) {
    __extends(FocusTrap, _super);
    function FocusTrap() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rootRef = createRef();
        return _this;
    }
    FocusTrap.prototype.componentDidMount = function () {
        var rootElement = this.rootRef.current;
        if (rootElement) {
            var trap = createFocusTrap(rootElement, {
                escapeDeactivates: false,
                fallbackFocus: rootElement
            });
            if (this.props.active) {
                trap.activate();
            }
            this.trap = trap;
        }
    };
    FocusTrap.prototype.componentWillUnmount = function () {
        if (this.trap) {
            this.trap.deactivate();
        }
    };
    FocusTrap.prototype.componentDidUpdate = function (prevProps) {
        if (this.trap && prevProps.active !== this.props.active) {
            this.props.active ? this.trap.activate() : this.trap.deactivate();
        }
    };
    FocusTrap.prototype.render = function () {
        var _a = this.props, active = _a.active, restProps = __rest(_a, ["active"]);
        return createElement("div", __assign({ ref: this.rootRef, tabIndex: -1 }, restProps));
    };
    return FocusTrap;
}(Component));

export { Board, EmptyMessage, FocusTrap };
