// @kata-kit/avatar/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var styled = require('styled-components');
var styled__default = _interopDefault(styled);
var ThemedComponent = require('@kata-kit/theme');
var ThemedComponent__default = _interopDefault(ThemedComponent);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var defaultTheme = {
    backgroundColor: ThemedComponent.variables.colors.gray10,
    textColor: ThemedComponent.variables.colors.gray80,
    borderColor: ThemedComponent.variables.colors.gray30
};
var themes = {
    defaultTheme: defaultTheme
};

var styles = function (props) {
    return styled.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    display: inline-block;\n    position: relative;\n    width: ", "px;\n    height: ", "px;\n    border: 1px solid ", ";\n    border-radius: 50%;\n    background: ", ";\n    font-size: 1rem;\n    line-height: 1;\n    vertical-align: middle;\n    overflow: hidden;\n\n    img {\n      position: absolute;\n      width: 100%;\n      height: 100%;\n      object-fit: cover;\n    }\n  "], ["\n    display: inline-block;\n    position: relative;\n    width: ", "px;\n    height: ", "px;\n    border: 1px solid ", ";\n    border-radius: 50%;\n    background: ", ";\n    font-size: 1rem;\n    line-height: 1;\n    vertical-align: middle;\n    overflow: hidden;\n\n    img {\n      position: absolute;\n      width: 100%;\n      height: 100%;\n      object-fit: cover;\n    }\n  "])), props.size, props.size, props.theme.borderColor, props.theme.backgroundColor);
};
var templateObject_1;

var Avatar = (function (_super) {
    __extends(Avatar, _super);
    function Avatar() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Avatar.prototype.render = function () {
        var _a = this.props, src = _a.src, size = _a.size, className = _a.className, style = _a.style;
        return (React.createElement(ThemedComponent__default, { themes: themes }, function (themeAttributes) { return (React.createElement(Root, { className: className, style: style, theme: themeAttributes, size: size },
            React.createElement(Image, { src: src }))); }));
    };
    Avatar.defaultProps = {
        size: 40
    };
    return Avatar;
}(React.Component));
var Image = styled__default('img')(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject([""], [""])));
var Root = styled__default('div')(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", ";\n"], ["\n  ", ";\n"])), styles);
var templateObject_1$1, templateObject_2;

exports.Avatar = Avatar;
