// tslint-config-kata/build/config/baseConfig.js
"use strict";
// Base Config
// Overrides default TSLint configs according to Kata Platform Development Standard.
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    rules: {
        // extended rules for tslint-config-airbnb + kata exceptions
        align: [true, 'statements'],
        ban: false,
        eofline: false,
        forin: true,
        'import-name': false,
        'interface-name': [true, 'always-prefix'],
        'jsdoc-format': true,
        'label-position': true,
        'max-line-length': [true, 120],
        'member-ordering': [true, { order: 'fields-first' }],
        'no-any': false,
        'no-arg': true,
        'no-bitwise': true,
        'no-conditional-assignment': true,
        'no-console': [true, 'log', 'error', 'debug', 'info', 'time', 'timeEnd', 'trace'],
        'no-debugger': true,
        'no-duplicate-variable': true,
        'no-dynamic-delete': true,
        'no-empty': true,
        'no-shadowed-variable': true,
        'no-string-literal': false,
        'no-switch-case-fall-through': true,
        'no-unused-expression': true,
        'no-use-before-declare': false,
        'one-line': [true, 'check-catch', 'check-else', 'check-open-brace', 'check-whitespace'],
        semicolon: [true, 'always', 'ignore-bound-class-methods'],
        'switch-default': true,
        'ter-indent': [true, 4, { SwitchCase: 1 }],
        typedef: [true, 'parameter', 'property-declaration'],
        'typedef-whitespace': [
            true,
            {
                'call-signature': 'nospace',
                'index-signature': 'nospace',
                parameter: 'nospace',
                'property-declaration': 'nospace',
                'variable-declaration': 'nospace',
            },
        ],
        'variable-name': [
            // 23.2 addition
            true,
            'ban-keywords',
            'check-format',
            'allow-leading-underscore',
            'allow-pascal-case',
        ],
    },
};
