// @kata-kit/dashboard/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var layout = require('@kata-kit/layout');
var tooltip = require('@kata-kit/tooltip');
var classNames = _interopDefault(require('classnames'));
var React = require('react');
var styled = require('styled-components');
var styled__default = _interopDefault(styled);
var theme = require('@kata-kit/theme');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}

function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var Dashboard = (function (_super) {
    __extends(Dashboard, _super);
    function Dashboard(props) {
        var _this = _super.call(this, props) || this;
        _this.renderStarter = _this.renderStarter.bind(_this);
        return _this;
    }
    Dashboard.prototype.renderStarter = function () {
        var _a = this.props, className = _a.className, title = _a.title, tooltip$$1 = _a.tooltip, subTitle = _a.subTitle, headerContent = _a.headerContent, image = _a.image, children = _a.children, isStarter = _a.isStarter;
        return (React.createElement(theme.Theme, null, function (themeAttributes) { return (React.createElement(Root, __assign({ className: classNames(className, isStarter && 'is-starter') }, themeAttributes),
            React.createElement(DashboardStarterHeading, null,
                React.createElement(layout.Container, null,
                    React.createElement(DashboardStarterInner, null,
                        React.createElement(DashboardStarterLeft, null,
                            React.createElement(DashboardHeader, null,
                                React.createElement(DashboardContainer, null,
                                    React.createElement(DashboardTitle, { isStarter: true, "data-testid": "Dashboard-title" }, title),
                                    tooltip$$1 && (React.createElement(tooltip.TooltipTarget, { component: React.createElement(tooltip.Tooltip, null, tooltip$$1) },
                                        React.createElement(DashboardTooltip, { className: "icon-info" }))))),
                            subTitle && (React.createElement(DashboardSubtitle, { isStarter: true }, subTitle)),
                            headerContent && (React.createElement(DashboardHeaderContent, null, headerContent))),
                        image && (React.createElement(DashboardStarterImage, null,
                            React.createElement("img", { src: image, alt: "header-img" })))))),
            React.createElement(DashboardContent, null,
                React.createElement(layout.Container, { "data-testid": "Dashboard-content" }, children)))); }));
    };
    Dashboard.prototype.render = function () {
        var _a = this.props, className = _a.className, title = _a.title, tooltip$$1 = _a.tooltip, subTitle = _a.subTitle, paragraph = _a.paragraph, headerContent = _a.headerContent, children = _a.children, isStarter = _a.isStarter;
        if (isStarter) {
            return this.renderStarter();
        }
        return (React.createElement(theme.Theme, null, function (themeAttributes) { return (React.createElement(Root, __assign({ className: className }, themeAttributes),
            React.createElement(layout.Container, null,
                React.createElement(DashboardHeader, null, headerContent || (React.createElement(DashboardContainer, null,
                    React.createElement(DashboardTitle, { "data-testid": "Dashboard-title" }, title),
                    tooltip$$1 && (React.createElement(tooltip.TooltipTarget, { trigger: "hover", component: React.createElement(tooltip.Tooltip, null, tooltip$$1) },
                        React.createElement(DashboardTooltip, { className: "icon-info" })))))),
                subTitle && React.createElement(DashboardSubtitle, null, subTitle),
                paragraph && (React.createElement(DashboardParagraph, { className: classNames(isStarter && 'is-starter') }, paragraph)),
                React.createElement(DashboardContent, { "data-testid": "Dashboard-content" }, children)))); }));
    };
    Dashboard.defaultProps = {
        isStarter: false
    };
    return Dashboard;
}(React.Component));
var DashboardIn = styled.keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"], ["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"])));
var DashboardContainer = styled__default(layout.Container)(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
var DashboardHeader = styled__default('div')(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", " {\n    display: flex;\n  }\n"], ["\n  ", " {\n    display: flex;\n  }\n"])), DashboardContainer);
var DashboardTooltip = styled__default('i')(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  flex: 0 0 auto;\n  align-self: center;\n  font-size: 20px;\n  color: #676b6d /* $gray-60 */;\n  margin-left: 0.615384615rem /* $space-1 */;\n\n  &::before {\n    vertical-align: middle;\n  }\n\n  &:hover {\n    color: black;\n  }\n"], ["\n  flex: 0 0 auto;\n  align-self: center;\n  font-size: 20px;\n  color: #676b6d /* $gray-60 */;\n  margin-left: 0.615384615rem /* $space-1 */;\n\n  &::before {\n    vertical-align: middle;\n  }\n\n  &:hover {\n    color: black;\n  }\n"])));
var DashboardTitle = styled__default('h1')(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  margin-bottom: 0;\n  flex: 0 0 auto;\n  color: ", ";\n"], ["\n  margin-bottom: 0;\n  flex: 0 0 auto;\n  color: ", ";\n"])), function (props) { return props.isStarter && 'white'; });
var DashboardSubtitle = styled__default('h2')(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  color: ", ";\n"], ["\n  color: ", ";\n"])), function (props) { return props.isStarter && 'white'; });
var DashboardParagraph = styled__default('p')(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"], ["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"])));
var DashboardStarterInner = styled__default('div')(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: column;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    flex-direction: row;\n    align-items: center;\n    justify-content: space-between;\n  }\n"], ["\n  display: flex;\n  flex-direction: column;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    flex-direction: row;\n    align-items: center;\n    justify-content: space-between;\n  }\n"])));
var DashboardStarterLeft = styled__default('div')(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  width: 100%;\n  max-width: 480px;\n"], ["\n  width: 100%;\n  max-width: 480px;\n"])));
var DashboardStarterImage = styled__default('div')(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  display: none;\n  height: 100%;\n  max-height: 184px;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    display: block;\n  }\n"], ["\n  display: none;\n  height: 100%;\n  max-height: 184px;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    display: block;\n  }\n"])));
var DashboardHeaderContent = styled__default('div')(templateObject_11 || (templateObject_11 = __makeTemplateObject(["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n\n  & p {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"], ["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n\n  & p {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"])));
var DashboardContent = styled__default('div')(templateObject_12 || (templateObject_12 = __makeTemplateObject(["\n  padding-top: ", ";\n"], ["\n  padding-top: ", ";\n"])), theme.variables.spaces.space3);
var DashboardStarterHeading = styled__default('div')(templateObject_13 || (templateObject_13 = __makeTemplateObject(["\n  display: flex;\n  padding: 40px 20px 136px;\n  background-color: #006fe6 /* $kata-blue */;\n\n  @media (min-width: 1280px /* $break-medium */) {\n    padding: 40px 40px 136px;\n  }\n\n  @media (min-width: 1366px /* $break-large */) {\n    padding: 48px 48px 136px;\n  }\n"], ["\n  display: flex;\n  padding: 40px 20px 136px;\n  background-color: #006fe6 /* $kata-blue */;\n\n  @media (min-width: 1280px /* $break-medium */) {\n    padding: 40px 40px 136px;\n  }\n\n  @media (min-width: 1366px /* $break-large */) {\n    padding: 48px 48px 136px;\n  }\n"])));
var Root = styled__default('div')(templateObject_14 || (templateObject_14 = __makeTemplateObject(["\n  animation: ", " 0.5s ease;\n\n  &:not(.is-starter) {\n    padding: 1.846153846rem /* $space-3 */ 20px;\n\n    @media (min-width: 1280px /* $break-medium */) {\n      padding: 1.846153846rem /* $space-3 */ 40px;\n    }\n\n    @media (min-width: 1366px /* $break-large */) {\n      padding: 1.846153846rem /* $space-3 */ 48px;\n    }\n  }\n\n  &.is-starter {\n    ", " {\n      h1 {\n        color: #fff;\n      }\n    }\n\n    ", " {\n      margin-top: -100px !important;\n      padding-top: 0 !important;\n    }\n  }\n"], ["\n  animation: ", " 0.5s ease;\n\n  &:not(.is-starter) {\n    padding: 1.846153846rem /* $space-3 */ 20px;\n\n    @media (min-width: 1280px /* $break-medium */) {\n      padding: 1.846153846rem /* $space-3 */ 40px;\n    }\n\n    @media (min-width: 1366px /* $break-large */) {\n      padding: 1.846153846rem /* $space-3 */ 48px;\n    }\n  }\n\n  &.is-starter {\n    ", " {\n      h1 {\n        color: #fff;\n      }\n    }\n\n    ", " {\n      margin-top: -100px !important;\n      padding-top: 0 !important;\n    }\n  }\n"])), DashboardIn, DashboardHeader, DashboardContent);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13, templateObject_14;

var DashboardCards = (function (_super) {
    __extends(DashboardCards, _super);
    function DashboardCards() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DashboardCards.prototype.render = function () {
        var _a = this.props, className = _a.className, cardsPerRow = _a.cardsPerRow, rest = __rest(_a, ["className", "cardsPerRow"]);
        return (React.createElement(Wrapper, __assign({ className: classNames(className) }, rest), React.Children.map(this.props.children, function (Item) {
            return Item ? (React.createElement(CardWrapper, { "data-testid": "CardGrid-wrapper", cardsPerRow: cardsPerRow }, React.cloneElement(Item, {
                dashboardCard: true
            }))) : (Item);
        })));
    };
    DashboardCards.defaultProps = {
        cardsPerRow: 3
    };
    return DashboardCards;
}(React.Component));
var Wrapper = styled__default('div')(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  display: flex;\n  flex-wrap: wrap;\n  margin: 0 -0.615384615rem;\n  justify-content: ", ";\n"], ["\n  display: flex;\n  flex-wrap: wrap;\n  margin: 0 -0.615384615rem;\n  justify-content: ", ";\n"])), function (props) { return props.centered && 'center'; });
var CardWrapper = styled__default('div')(templateObject_2$1 || (templateObject_2$1 = __makeTemplateObject(["\n  flex: 0 1 100%;\n  padding: 0 0.615384615rem;\n  margin-bottom: 1.230769231rem;\n\n  @media (min-width: ", ") {\n    flex: 0 1 50%;\n  }\n\n  @media (min-width: ", ") {\n    flex: ", ";\n  }\n"], ["\n  flex: 0 1 100%;\n  padding: 0 0.615384615rem;\n  margin-bottom: 1.230769231rem;\n\n  @media (min-width: ", ") {\n    flex: 0 1 50%;\n  }\n\n  @media (min-width: ", ") {\n    flex: ", ";\n  }\n"])), theme.variables.breaks.breakSmall, theme.variables.breaks.breakMedium, function (props) { return (props.cardsPerRow === 3 ? '0 1 33.33%' : '0 1 50%'); });
var templateObject_1$1, templateObject_2$1;

var DashboardContentHeader = function (_a) {
    var children = _a.children, className = _a.className, isSecondary = _a.isSecondary;
    return (React.createElement(theme.Theme, null, function (themeAttributes) { return (React.createElement(Root$1, __assign({ className: className, isSecondary: isSecondary }, themeAttributes), children)); }));
};
var Root$1 = styled__default('h1')(templateObject_1$2 || (templateObject_1$2 = __makeTemplateObject(["\n  margin-top: ", ";\n  margin-bottom: 24px;\n  color: ", ";\n  font-size: 1.85rem;\n  font-weight: 500;\n"], ["\n  margin-top: ", ";\n  margin-bottom: 24px;\n  color: ",
    ";\n  font-size: 1.85rem;\n  font-weight: 500;\n"])), function (props) { return (props.isSecondary ? '2.461538462rem' : 0); }, function (props) {
    return props.isSecondary ? '#24282d !important' : '#ffffff !important';
});
var templateObject_1$2;

exports.Dashboard = Dashboard;
exports.DashboardCards = DashboardCards;
exports.DashboardContentHeader = DashboardContentHeader;
