// @kata-kit/dashboard/lib/index.es.js
import { Container } from '@kata-kit/layout';
import { Tooltip, TooltipTarget } from '@kata-kit/tooltip';
import classNames from 'classnames';
import { createElement, Component, Children, cloneElement } from 'react';
import styled, { keyframes } from 'styled-components';
import { Theme, variables } from '@kata-kit/theme';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}

function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var Dashboard = (function (_super) {
    __extends(Dashboard, _super);
    function Dashboard(props) {
        var _this = _super.call(this, props) || this;
        _this.renderStarter = _this.renderStarter.bind(_this);
        return _this;
    }
    Dashboard.prototype.renderStarter = function () {
        var _a = this.props, className = _a.className, title = _a.title, tooltip = _a.tooltip, subTitle = _a.subTitle, headerContent = _a.headerContent, image = _a.image, children = _a.children, isStarter = _a.isStarter;
        return (createElement(Theme, null, function (themeAttributes) { return (createElement(Root, __assign({ className: classNames(className, isStarter && 'is-starter') }, themeAttributes),
            createElement(DashboardStarterHeading, null,
                createElement(Container, null,
                    createElement(DashboardStarterInner, null,
                        createElement(DashboardStarterLeft, null,
                            createElement(DashboardHeader, null,
                                createElement(DashboardContainer, null,
                                    createElement(DashboardTitle, { isStarter: true, "data-testid": "Dashboard-title" }, title),
                                    tooltip && (createElement(TooltipTarget, { component: createElement(Tooltip, null, tooltip) },
                                        createElement(DashboardTooltip, { className: "icon-info" }))))),
                            subTitle && (createElement(DashboardSubtitle, { isStarter: true }, subTitle)),
                            headerContent && (createElement(DashboardHeaderContent, null, headerContent))),
                        image && (createElement(DashboardStarterImage, null,
                            createElement("img", { src: image, alt: "header-img" })))))),
            createElement(DashboardContent, null,
                createElement(Container, { "data-testid": "Dashboard-content" }, children)))); }));
    };
    Dashboard.prototype.render = function () {
        var _a = this.props, className = _a.className, title = _a.title, tooltip = _a.tooltip, subTitle = _a.subTitle, paragraph = _a.paragraph, headerContent = _a.headerContent, children = _a.children, isStarter = _a.isStarter;
        if (isStarter) {
            return this.renderStarter();
        }
        return (createElement(Theme, null, function (themeAttributes) { return (createElement(Root, __assign({ className: className }, themeAttributes),
            createElement(Container, null,
                createElement(DashboardHeader, null, headerContent || (createElement(DashboardContainer, null,
                    createElement(DashboardTitle, { "data-testid": "Dashboard-title" }, title),
                    tooltip && (createElement(TooltipTarget, { trigger: "hover", component: createElement(Tooltip, null, tooltip) },
                        createElement(DashboardTooltip, { className: "icon-info" })))))),
                subTitle && createElement(DashboardSubtitle, null, subTitle),
                paragraph && (createElement(DashboardParagraph, { className: classNames(isStarter && 'is-starter') }, paragraph)),
                createElement(DashboardContent, { "data-testid": "Dashboard-content" }, children)))); }));
    };
    Dashboard.defaultProps = {
        isStarter: false
    };
    return Dashboard;
}(Component));
var DashboardIn = keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"], ["\n  from {\n    opacity: 0;\n  }\n  to {\n    opacity: 1;\n  }\n"])));
var DashboardContainer = styled(Container)(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
var DashboardHeader = styled('div')(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", " {\n    display: flex;\n  }\n"], ["\n  ", " {\n    display: flex;\n  }\n"])), DashboardContainer);
var DashboardTooltip = styled('i')(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  flex: 0 0 auto;\n  align-self: center;\n  font-size: 20px;\n  color: #676b6d /* $gray-60 */;\n  margin-left: 0.615384615rem /* $space-1 */;\n\n  &::before {\n    vertical-align: middle;\n  }\n\n  &:hover {\n    color: black;\n  }\n"], ["\n  flex: 0 0 auto;\n  align-self: center;\n  font-size: 20px;\n  color: #676b6d /* $gray-60 */;\n  margin-left: 0.615384615rem /* $space-1 */;\n\n  &::before {\n    vertical-align: middle;\n  }\n\n  &:hover {\n    color: black;\n  }\n"])));
var DashboardTitle = styled('h1')(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  margin-bottom: 0;\n  flex: 0 0 auto;\n  color: ", ";\n"], ["\n  margin-bottom: 0;\n  flex: 0 0 auto;\n  color: ", ";\n"])), function (props) { return props.isStarter && 'white'; });
var DashboardSubtitle = styled('h2')(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  color: ", ";\n"], ["\n  color: ", ";\n"])), function (props) { return props.isStarter && 'white'; });
var DashboardParagraph = styled('p')(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"], ["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"])));
var DashboardStarterInner = styled('div')(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: column;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    flex-direction: row;\n    align-items: center;\n    justify-content: space-between;\n  }\n"], ["\n  display: flex;\n  flex-direction: column;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    flex-direction: row;\n    align-items: center;\n    justify-content: space-between;\n  }\n"])));
var DashboardStarterLeft = styled('div')(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  width: 100%;\n  max-width: 480px;\n"], ["\n  width: 100%;\n  max-width: 480px;\n"])));
var DashboardStarterImage = styled('div')(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  display: none;\n  height: 100%;\n  max-height: 184px;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    display: block;\n  }\n"], ["\n  display: none;\n  height: 100%;\n  max-height: 184px;\n\n  @media (min-width: 1024px /* $break-medium */) {\n    display: block;\n  }\n"])));
var DashboardHeaderContent = styled('div')(templateObject_11 || (templateObject_11 = __makeTemplateObject(["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n\n  & p {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"], ["\n  &:not(.is-starter) {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n\n  & p {\n    margin-top: 8px;\n    max-width: 480px;\n    word-wrap: break-word;\n    color: #fff !important;\n    font-weight: 300;\n  }\n"])));
var DashboardContent = styled('div')(templateObject_12 || (templateObject_12 = __makeTemplateObject(["\n  padding-top: ", ";\n"], ["\n  padding-top: ", ";\n"])), variables.spaces.space3);
var DashboardStarterHeading = styled('div')(templateObject_13 || (templateObject_13 = __makeTemplateObject(["\n  display: flex;\n  padding: 40px 20px 136px;\n  background-color: #006fe6 /* $kata-blue */;\n\n  @media (min-width: 1280px /* $break-medium */) {\n    padding: 40px 40px 136px;\n  }\n\n  @media (min-width: 1366px /* $break-large */) {\n    padding: 48px 48px 136px;\n  }\n"], ["\n  display: flex;\n  padding: 40px 20px 136px;\n  background-color: #006fe6 /* $kata-blue */;\n\n  @media (min-width: 1280px /* $break-medium */) {\n    padding: 40px 40px 136px;\n  }\n\n  @media (min-width: 1366px /* $break-large */) {\n    padding: 48px 48px 136px;\n  }\n"])));
var Root = styled('div')(templateObject_14 || (templateObject_14 = __makeTemplateObject(["\n  animation: ", " 0.5s ease;\n\n  &:not(.is-starter) {\n    padding: 1.846153846rem /* $space-3 */ 20px;\n\n    @media (min-width: 1280px /* $break-medium */) {\n      padding: 1.846153846rem /* $space-3 */ 40px;\n    }\n\n    @media (min-width: 1366px /* $break-large */) {\n      padding: 1.846153846rem /* $space-3 */ 48px;\n    }\n  }\n\n  &.is-starter {\n    ", " {\n      h1 {\n        color: #fff;\n      }\n    }\n\n    ", " {\n      margin-top: -100px !important;\n      padding-top: 0 !important;\n    }\n  }\n"], ["\n  animation: ", " 0.5s ease;\n\n  &:not(.is-starter) {\n    padding: 1.846153846rem /* $space-3 */ 20px;\n\n    @media (min-width: 1280px /* $break-medium */) {\n      padding: 1.846153846rem /* $space-3 */ 40px;\n    }\n\n    @media (min-width: 1366px /* $break-large */) {\n      padding: 1.846153846rem /* $space-3 */ 48px;\n    }\n  }\n\n  &.is-starter {\n    ", " {\n      h1 {\n        color: #fff;\n      }\n    }\n\n    ", " {\n      margin-top: -100px !important;\n      padding-top: 0 !important;\n    }\n  }\n"])), DashboardIn, DashboardHeader, DashboardContent);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13, templateObject_14;

var DashboardCards = (function (_super) {
    __extends(DashboardCards, _super);
    function DashboardCards() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DashboardCards.prototype.render = function () {
        var _a = this.props, className = _a.className, cardsPerRow = _a.cardsPerRow, rest = __rest(_a, ["className", "cardsPerRow"]);
        return (createElement(Wrapper, __assign({ className: classNames(className) }, rest), Children.map(this.props.children, function (Item) {
            return Item ? (createElement(CardWrapper, { "data-testid": "CardGrid-wrapper", cardsPerRow: cardsPerRow }, cloneElement(Item, {
                dashboardCard: true
            }))) : (Item);
        })));
    };
    DashboardCards.defaultProps = {
        cardsPerRow: 3
    };
    return DashboardCards;
}(Component));
var Wrapper = styled('div')(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  display: flex;\n  flex-wrap: wrap;\n  margin: 0 -0.615384615rem;\n  justify-content: ", ";\n"], ["\n  display: flex;\n  flex-wrap: wrap;\n  margin: 0 -0.615384615rem;\n  justify-content: ", ";\n"])), function (props) { return props.centered && 'center'; });
var CardWrapper = styled('div')(templateObject_2$1 || (templateObject_2$1 = __makeTemplateObject(["\n  flex: 0 1 100%;\n  padding: 0 0.615384615rem;\n  margin-bottom: 1.230769231rem;\n\n  @media (min-width: ", ") {\n    flex: 0 1 50%;\n  }\n\n  @media (min-width: ", ") {\n    flex: ", ";\n  }\n"], ["\n  flex: 0 1 100%;\n  padding: 0 0.615384615rem;\n  margin-bottom: 1.230769231rem;\n\n  @media (min-width: ", ") {\n    flex: 0 1 50%;\n  }\n\n  @media (min-width: ", ") {\n    flex: ", ";\n  }\n"])), variables.breaks.breakSmall, variables.breaks.breakMedium, function (props) { return (props.cardsPerRow === 3 ? '0 1 33.33%' : '0 1 50%'); });
var templateObject_1$1, templateObject_2$1;

var DashboardContentHeader = function (_a) {
    var children = _a.children, className = _a.className, isSecondary = _a.isSecondary;
    return (createElement(Theme, null, function (themeAttributes) { return (createElement(Root$1, __assign({ className: className, isSecondary: isSecondary }, themeAttributes), children)); }));
};
var Root$1 = styled('h1')(templateObject_1$2 || (templateObject_1$2 = __makeTemplateObject(["\n  margin-top: ", ";\n  margin-bottom: 24px;\n  color: ", ";\n  font-size: 1.85rem;\n  font-weight: 500;\n"], ["\n  margin-top: ", ";\n  margin-bottom: 24px;\n  color: ",
    ";\n  font-size: 1.85rem;\n  font-weight: 500;\n"])), function (props) { return (props.isSecondary ? '2.461538462rem' : 0); }, function (props) {
    return props.isSecondary ? '#24282d !important' : '#ffffff !important';
});
var templateObject_1$2;

export { Dashboard, DashboardCards, DashboardContentHeader };
